Allow some tolerance in 3D qp projection

lindsayad opened the pull request:
Issue reported in #23925

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1442188/) on e9f1f68 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23954/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1442202/) on e9f1f68 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>40cceb</th>
      <th colspan=4>#23954 e9f1f6</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.91%</td>
      <td>84.91%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>84351</td>
      <td>84350</td>
      <td>-1</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14987</td>
      <td>14990</td>
      <td>+3</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23954/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23954/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23954/coverage/xfem)

</details>



This comment will be updated on new commits.

recuero commented on framework/src/utils/MortarUtils.C:
```diff
@@ -262,13 +262,15 @@ projectQPoints3d(const Elem * const msm_elem,
-        if (qp_back(0) < 0 || qp_back(1) < 0 || qp_back(0) + qp_back(1) > 1)
+        if (qp_back(0) < -TOLERANCE || qp_back(1) < -TOLERANCE ||
+            qp_back(0) + qp_back(1) > (1 + TOLERANCE))
           mooseException("Quadrature point: ", qp_back, " out of bounds, truncating.");
```
Can the quadrature points be forced back on the feasible range if we fall within `TOLERANCE` range?

lindsayad replied:
Hmm it's immediately obvious what to do for the 1D comparisons but what about for the 2D comparison ... perhaps just setting it to `qp_back.unit()`

lindsayad replied:
I guess my opinion is that if none our later code barfs on these quadrature points within tolerance of the reference domain (but strictly speaking outside of it), that I think it's fine as is

lindsayad replied:
@bozeng1997's thermomechanical simulation is pretty neat. With the commit here allowing it to run, here's at the beginning of the sim

<img width="1324" alt="Screen Shot 2023-04-06 at 3 21 03 PM" src="https://user-images.githubusercontent.com/10248304/230504419-57cc23b0-87db-4f06-8ecc-dbb60173104e.PNG">
 
and here's after 100 time steps

<img width="1321" alt="Screen Shot 2023-04-06 at 3 21 14 PM" src="https://user-images.githubusercontent.com/10248304/230504450-83056c9c-d1cc-4a5c-af2a-0d339a30dc30.PNG">

lindsayad replied:
Regardless of whether moving the qps back into the strict reference domain or not, I should add a test. @bozeng97 could I use your input file for the test I add to the MOOSE regression suite?

recuero replied:
I am fine with the code as is too. I do feel like there should be a better way to tackle the variability in problem units, but that's a story for a different day.

BoZeng1997 replied:
@lindsayad sure. please use the input file I provided. I am glad that I can help.

lindsayad replied:
great, thanks!

lindsayad committed:
Test for slightly outside 3d mortar projections

lindsayad committed:
Allow some tolerance in 3D qp projection

Refs #23925 #24032

moosebuild commented:
Job [Modules](https://civet.inl.gov/job/1455577/) on e2600d7 : invalidated by @lindsayad

want to ensure new test runs

recuero commented on modules/combined/test/tests/3d-mortar-projection-tolerancing/test.i:
```diff
@@ -0,0 +1,228 @@
+    cylinder_axis_point_1 = '0 0 0'
+    cylinder_axis_point_2 = '0 0 1'
+    use_displaced_mesh = true
+    quadrature = SECOND
```
What's this parameter doing? I don't recall it.

recuero replied:
It may be a remnant of the `quadrature = true` option from the older boundary condition.

BoZeng1997 replied:
I set `quadrature=SECOND` because the description provided with this parameter
```
Default:DEFAULT
C++ Type:MooseEnum
Options:DEFAULT, FIRST, SECOND, THIRD, FOURTH
Controllable:No
Description:Quadrature rule to use on mortar segments. For 2D mortar DEFAULT is recommended. For 3D mortar, QUAD meshes are integrated using triangle mortar segments. While DEFAULT quadrature order is typically sufficiently accurate, exact integration of QUAD mortar faces requires SECOND order quadrature for FIRST variables and FOURTH order quadrature for SECOND order variables.
```

recuero replied:
It seems to me the `DEFAULT` should pick an integration rule according to the mesh. For a problem with first order elements, it would be 2 * LINEAR + 1 = 3. So `DEFAULT` may integrate with more points. Thanks for pointing that out.

recuero replied:
I think we should probably automate the selection of the integration rule and discourage (not forbid) controlling that parameter from the input file.

recuero commented on modules/combined/test/tests/3d-mortar-projection-tolerancing/test.i:
```diff
@@ -0,0 +1,228 @@
+  petsc_options_value = 'lu      superlu_dist NONZERO 1e-14'
+  automatic_scaling = true
+  line_search = none
+  ignore_variables_for_autoscaling = 'pellet_normal_lm  lm_pellet '
```
I understand that ignoring the scaling of the mechanical contact LM can help due to its high nonlinearity and that mechanical contact typically takes place suddenly. How about the thermal LM: Can scaling be applied to it?

recuero replied:
Just for my edification.

lindsayad replied:
For on-diagonal automatic scaling which is the default, scaling the mechanical LM can be problematic as sometimes there is no on-diagonal. We should be fine to auto-scale with the diagonal of the thermal LM since it does have an on-diagonal. I'll try making that change

BoZeng1997 replied:
When scaling thermal LMs, do we need to explicitly specify thermal LMs having their scaling factor computed separately like `scaling_group_variables = 'disp_x disp_y disp_z; T_K; LM1; LM2'` ? Is `scaling_group_variables = 'disp_x disp_y disp_z; T_K'` a **safe** way to let MOOSE knows that we want to compute the LMs variables's scaling factor separately?

lindsayad replied:
I had to double check to be sure, but yes it appears that scaling will apply safely to variables not included in the `scaling_group_variables` list. So if pass `Executioner/verbose=true` to the input you gave me, I see this for the scaling factors:
```
Automatic scaling factors:
  disp_x: 2.94915e-09
  disp_y: 2.94915e-09
  disp_z: 2.94915e-09
  T_K: 17.3236
  lm_pellet: 6.51664e+07
  pellet_normal_lm: 1
```

recuero replied:
> For on-diagonal automatic scaling which is the default, scaling the mechanical LM can be problematic as sometimes there is no on-diagonal. We should be fine to auto-scale with the diagonal of the thermal LM since it does have an on-diagonal. I'll try making that change

I believe scaling happens at each time step (at most). I have had cases where, despite disabling the `compute_only_once` option and using off diagonal terms, if mechanical contact took place in the second or third nonlinear iteration, the simulation would not be able to continue. I had to disable automatic scaling on the mechanical contact LM in some of those cases.

recuero commented on modules/combined/test/tests/3d-mortar-projection-tolerancing/test.i:
```diff
@@ -0,0 +1,228 @@
+  []
+  [lm_pellet]
+    block = 'pellet_secondary_subdomain'
+    use_dual = true
```
I think we don't need dual bases to enforce the thermal problem.

recuero commented on modules/combined/test/tests/3d-mortar-projection-tolerancing/test.i:
```diff
@@ -0,0 +1,228 @@
+    primary = void_pellet_0
+    secondary = void_pellet_1
+    model = frictionless
+    penalty = 1e6
```
This penalty parameter is unused.

lindsayad committed:
Clean up some spacing

lindsayad committed:
Address Antonio review

- Don't use dual for thermal LM bc it naturally has a diagonal
- For the same reason, we don't have to ignore it in auto scaling
- Remove unused penalty parameter in contact action

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1459111/) on 7ff120e wanted to post the following:


The following coverage requirement(s) failed:
- `misc` coverage rate 34.31% is less than the required 55.0%
- `phase_field` coverage rate 86.27% is less than the required 89.0%
- `richards` coverage rate 94.75% is less than the required 97.0%
- `tensor_mechanics` coverage rate 85.21% is less than the required 86.0%
